packages/kit/src/core/sync/write_types/index.js

```diff
@@ -155,48 +171,44 @@ function update_types(config, manifest_data, route) {
 		declarations.push(`interface RouteParams extends Partial<Record<string, string>> {}`);
 	}
 
-	if (route.leaf) {
-		const { data, server_data, load, server_load, errors, written_proxies } = process_node(
-			route.leaf,
-			outdir,
-			'RouteParams'
+	if (route.layout || route.leaf) {
+		// These could also be placed in our public types, but it would bloat them unnecessarily and we may want to change these in the future
+		declarations.push(`type MaybeWithVoid<T> = {} extends T ? T | void : T;`);
+		declarations.push(
+			`export type RequiredKeys<T> = { [K in keyof T]-?: {} extends { [P in K]: T[K] } ? never : K; }[keyof T];`
 		);
+		declarations.push(
+			`type OutputDataShape<T> = MaybeWithVoid<Omit<App.PageData, RequiredKeys<T>> & Partial<Pick<App.PageData, keyof T & keyof App.PageData>> & Record<string, any>>`
+		);
+		declarations.push(`type EnsureParentData<T> = NonNullable<T> extends never ? {} : T;`);
+	}
 
-		for (const file of written_proxies) to_delete.delete(file);
-
-		exports.push(`export type Errors = ${errors};`);
+	if (route.leaf) {
+		const { declarations: d, exports: e, written_proxies } = process_node(route.leaf, outdir, true);
 
-		exports.push(`export type PageData = ${data};`);
-		if (load) {
-			exports.push(
-				`export type PageLoad<OutputData extends Record<string, any> | void = Record<string, any> | void> = ${load};`
-			);
-			exports.push('export type PageLoadEvent = Parameters<PageLoad>[0];');
-		}
+		exports.push(...e);
+		declarations.push(...d);
 
-		exports.push(`export type PageServerData = ${server_data};`);
-		if (server_load) {
-			exports.push(
-				`export type PageServerLoad<OutputData extends Record<string, any> | void = Record<string, any> | void> = ${server_load};`
-			);
-			exports.push('export type PageServerLoadEvent = Parameters<PageServerLoad>[0];');
-		}
+		for (const file of written_proxies) to_delete.delete(file);
 
 		if (route.leaf.server) {
 			exports.push(`export type Action = Kit.Action<RouteParams>`);
 		}
 	}
 
 	if (route.layout) {
-		// TODO collect children in create_manifest_data, instead of this inefficient O(n^2) algorithm
+		let all_pages_have_load = true;
 		const layout_params = new Set();
-		manifest_data.routes.forEach((other) => {
-			if (other.page && other.id.startsWith(route.id + '/')) {
-				// TODO this is O(n^2), see if we need to speed it up
-				for (const name of other.names) {
+		route.layout.child_pages?.forEach((page) => {
+			const leaf = routes.get(page);
+			if (leaf) {
+				for (const name of leaf.names) {
 					layout_params.add(name);
 				}
 			}
+			if (!page?.server && !page?.shared) {
```

Comment:
is the optional chaining necessary? haven't fully grokked the PR yet but at first glance it feels like `child_pages` shouldn't have any falsy values

```suggestion
			if (!page?.server && !page?.shared) {
```